### PR TITLE
feat(catalog-backend-module-gitlab): add option for custom location c…

### LINKED ISSUE #16859: 🚀 Feature: plugin-catalog-backend-module-gitlab support custom
### 🔖 Feature description

We use the Gitlab catalog discovery plugin, this works as defined. In our use case we want to use a specific `type` and `catalogFile`. The custom type is working fine if we import manually. The issue is that the Gitlab discovery import only supports the default `url` type. 

### 🎤 Context

- Support custom yaml file implementations for the gitlab discovery
- Not sure if other discovery plugins have similar needs

### ✌️ Possible Implementation

source file `src/providers/GitlabDiscoveryEntityProvider.ts``

```
  private createLocationSpec(project: GitLabProject): LocationSpec {
    const project_branch = project.default_branch ?? this.config.fallbackBranch;
    return {
      type: this.config.catalogType ?? 'url',
      target: `${project.web_url}/-/blob/${project_branch}/${this.config.catalogFile}`,
      presence: 'optional',
    };
  }
```

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by andreabettich**
…reation function

closes #16859

## Hey, I just made a Pull Request!

Add an option to the gitlab backend module, to customize the location creation. More details are described in the issue #16859 

Just note, I have almost no experience with typescript or js, so maybe there are some flaws in the code and I'm very open to feedback and suggestions. I'll add tests, when the code seems to be alright.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
Hey :wave: thanks for this! I think there's a few things you'll need to do to get this PR green!

You'll need to run `yarn install` which will probably change the `yarn.lock` so you'll need to commit those changes.

You'll also probably need to run `yarn build:api-reports` and commit the changes for those API reports too, as these help us see what the public API is for different packages, and that has also changed in this PR :tada:

**Comment by andreabettich** [comment]
Thanks for the instruction and I added the files. Got some warnings in the api report but figured it out but now I'm not sure if the type export is at the right place or not.

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts [hunk 1]
```diff
@@ -185,22 +192,19 @@ export class GitlabDiscoveryEntityProvider implements EntityProvider {
...
       );
       if (projectHasFile) {
         res.matches.push(project);
       }
     }
 
-    const locations = res.matches.map(p => this.createLocationSpec(p));
+    const locations = res.matches
+      .map(project => this.createLocationUrl(project))
+      .map(targetUrl => this.createLocationFn(targetUrl));
```
Comment (Edje-C) [inline plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts:207]: Here we have it iterating through our locations list twice, but I think the way it was can work by just calling `this. createLocationFn` in `createLocationSpec`

Comment (andreabettich) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts:207]: So you mean like this?
```
const locations = res.matches
      .map(project => this.createLocationFn(this.createLocationUrl(project)));
```

File: plugins/catalog-backend-module-gitlab/api-report.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ import { Config } from '@backstage/config';
 import { EntityProvider } from '@backstage/plugin-catalog-node';
 import { EntityProviderConnection } from '@backstage/plugin-catalog-node';
 import { LocationSpec } from '@backstage/plugin-catalog-node';
+import { LocationSpec as LocationSpec_2 } from '@backstage/plugin-catalog-common';
```
[the remark this answers is not in the document]
Comment (andreabettich) [reply plugins/catalog-backend-module-gitlab/api-report.md:12]: When I used the other one I got the warning or even an error that it's marked deprecated, thats why I used the one from `catalog-common`

File: plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts (hunk 1 above)
Comment (Edje-C) [reply plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts:207]: Like the way it was before
```suggestion
      const locations = res.matches.map(p => this.createLocationSpec(p));
```

But just

```
private createLocationSpec(project: GitLabProject): LocationSpec {
    const project_branch =
      this.config.branch ??
      project.default_branch ??
      this.config.fallbackBranch;
    
    return this.createLocationFn();
  }
```

**Pull request closed without merging**